packages/@aws-cdk/aws-s3/lib/notifications-resource/lambda-source/src/index.py

```diff
@@ -0,0 +1,84 @@
+import boto3  # type: ignore
+import json
+import urllib.request
+from typing import List
+
+s3 = boto3.client("s3")
+
+
+def handler(event: dict, context):
+    response_status = "SUCCESS"
+    error_message = ""
+    try:
+        props = event["ResourceProperties"]
+        bucket = props["BucketName"]
+        s3.put_bucket_notification_configuration(
+            Bucket=bucket,
+            NotificationConfiguration=get_config(
+                current_config=s3.get_bucket_notification_configuration(Bucket=bucket),
+                new_config=props["NotificationConfiguration"],
+                old_config=event.get("OldResourceProperties", {}).get("NotificationConfiguration", {}),
+                request_type=event["RequestType"],
+            ),
+        )
+    except Exception as e:
+        print("Failed to put bucket notification configuration:", e)
+        response_status = "FAILED"
+        error_message = f"Error: {str(e)}. "
+    finally:
+        submit_response(event, context, response_status, error_message)
+
+
+def get_config(current_config: dict, new_config: dict, old_config: dict, request_type: str) -> dict:
+    config: dict = {}
+    types = (
+        ("TopicConfigurations", "TopicArn"),
+        ("QueueConfigurations", "QueueArn"),
+        ("LambdaFunctionConfigurations", "LambdaFunctionArn"),
+    )
+    for config_type, target in types:
+        configs = current_config.get(config_type, [])
+        in_configs = new_config.get(config_type, [])
+        if request_type == "Create":
+            configs.extend(in_configs)
+            config[config_type] = configs
+        elif request_type == "Update":
+            configs = find_difference(configs, old_config.get(config_type, []), target)
+            configs.extend(find_difference(in_configs, configs, target))
+            config[config_type] = configs
+        elif request_type == "Delete":
+            config[config_type] = find_difference(configs, in_configs, target)
```

Comment:
The assumption here is that the incoming config in this scenario is the latest working configuration of the resource. This is true when called during stack deletion. However, when `DELETE` is called as part of a rollback, the properties are the actual incoming properties from the template. 

This means the following scenario is broken:

```ts
const bucket = s3.Bucket.fromBucketName(this, 'Bucket', props.existing.bucket.bucketName);
const topic = sns.Topic.fromTopicArn(this, 'Topic', props.existing.topic.topicArn);
const snsDestination = new notifications.SnsDestination(topic);

// if this notification already exists on the bucket - the deployment will fail (as it should). 
// Problem is that when the rollback completes, this notification will be deleted from the bucket.
bucket.addEventNotification(s3.EventType.OBJECT_CREATED_POST, snsDestination);
```

The only solution I see is to query CloudFormation and identify this scenario - i.e - if the event is `DELETE` and the stack is being rolledback, it means the original event was `CREATE` - which means we actually don't have to do anything.

Reply:
@iliapolo  any idea how to identify that kind of edge case.

Reply:
Use boto to query the stack state (you have the stack id in the CloudFormation event).

https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/cloudformation.html#CloudFormation.Client.describe_stacks

Reply:
Ok thanks, i will look it the stack is in the `ROLLBACK_IN_PROGRESS` state, and ignore the `Delete` request type.

Reply:
@iliapolo `StackName` is required, we only have the `StackId`.


```python3
response = client.describe_stacks(StackName='string')
```

Reply:
ok, I guess it supported both `The name or the unique stack ID that is associated with the stack`

Reply:
@iliapolo  ok I implemented this part see test case `test_delete_when_rollback_in_progress`